### PR TITLE
[SPARK-8233][SQL] misc function: hash

### PR DESCRIPTION
**Pull request by qiansl127**

### CODE REVIEW

**Comment by qiansl127** [comment]
@chenghao-intel @adrian-wang @zhichao-li

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+ */
+
+package org.apache.spark.sql.catalyst.expressions
+
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.types.{DataType, IntegerType}
+
+/**
+ * A function that returns a hash value of the argument
+ *
```
Comment (adrian-wang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: remove this line.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MiscFunctionsSuite.scala [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    checkEvaluation(Hash(Literal(3.7f)), 1080872141)
+    val v1: Byte = 3
+    val v2: Short = 3
+    checkEvaluation(Hash(Literal(v1)), 3)
+    checkEvaluation(Hash(Literal(v2)), 3)
+    checkEvaluation(Hash(Literal("ABC")), 94369)
+    checkEvaluation(Hash(Literal(true)), 1231)
+    checkEvaluation(Hash(Literal.create(3.7, DecimalType.Unlimited)), -644612093)
+    checkEvaluation(Hash(Literal.create(java.sql.Date.valueOf("1991-12-07"), DateType)), 8010)
+    checkEvaluation(Hash(Literal.create(java.sql.Timestamp.valueOf("1991-12-07 12:00:00"), TimestampType)), -1745111446)
```
Comment (adrian-wang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MiscFunctionsSuite.scala]: this line is too long.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+ */
+case class Hash(child: Expression) extends UnaryExpression {
```
Comment (adrian-wang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: hash should be able to handle an expression seq.

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 4]
```diff
@@ -1359,6 +1360,26 @@ object functions {
...
+  /**
+   * Calculates the hash value of the argument.
+   *
+   * @group misc_funcs
+   * @since 1.5.0
+   */
+  def hash(e: Column): Column = Hash(e.expr)
+
+  /**
+   * Calculates the hash value of the argument.
```
Comment (adrian-wang) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala]: column

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  override def eval(input: InternalRow): Any = {
+    val value = child.eval(input)
+    hashCode(value)
+  }
+
+  def hashCode(v: Any): Int = v match {
+    case null => 0
+    case arr: Array[_] => {
+      var res: Int = 0
+      arr.foreach { case(i) => res += hashCode(i) }
```
Comment (adrian-wang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: you could use foldLeft

Comment (adrian-wang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: or reduce.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+      var res: Int = 0
+      for (i <- 0 until r.length) {
+        res += 31 * res + hashCode(r.get(i))
+      }
+      res
+    }
+    case others => others.hashCode()
+  }
+
+  override def toString: String = s"hash($child)"
```
Comment (adrian-wang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/misc.scala]: May be use "HASH"?

**Pull request closed without merging**